### ISSUE TITLE
Proposal: Module Loader API

### ISSUE THREAD

**Issue by kitsonk** [denoland/deno#8327]
With ES Modules, it is effectively impossible to do "hot module reloading" of the modules, as ES Module specification is fairly strict that once the module is resolve and its two pass instantiation is done, that is the end of the road for the module and cannot be replaced.

When you are mocking/testing/etc. though, it is potentially valuable to make changes to a module and "reload" it via dynamic import.  In Node.js, they have solved this problem with an [experimental loader API](https://nodejs.org/api/esm.html#esm_experimental_loaders).  For an example of how this gets integrated to provide a mechanism for module replacements of imports that are still modules, [quibble](https://github.com/testdouble/quibble#es-modules-support) provides ESM support.  It basically creates a unique specifier for every time the module is requested to get around the challenge of ES modules being static once resolved.

This would also support solving things like #1739, which we have long wanted to do.

**Comment by chrisabrams** [comment]
@kitsonk did you ever come across a solution for `readTextFile` invalidation?

**Comment by happyhunter7** [comment]
So any news on this)

Deno team consistently closes every new task  opened and links to this
And no action on this for 4 years

If you check above at least 10issues closed in favor to this and nothing is happening

cc @lucacasonato just pointing to a big problem, especially, a bug closed with no reason https://github.com/denoland/deno/issues/25742

**Comment by bradthomasbrown** [comment]
![image](https://github.com/user-attachments/assets/443a80dc-06f0-4c48-8b09-e3a1c877b6e1)
It seems the Deno team not only closes the issues, doesn't offer much in terms of news or progress, but is also minimizing when this is pointed out.
That's incredibly disappointing, so far I have found Deno highly preferable as a very lightweight and powerful scripting/programming engine.

**Comment by petamoriken** [comment]
This issue is being discussed on the TC39 side.
https://github.com/tc39/proposal-esm-phase-imports
https://github.com/tc39/proposal-compartments

The following slide from the 8 October 2024 TC39 meeting is helpful.
https://docs.google.com/presentation/d/1HF4COMfypVzftilhOIlGxz9Fn1iaiFdwYDZwzqQt1gs/edit?usp=sharing

**Comment by gherciu** [comment]
The position of Deno team is a bit strange:
- It copies most of the Node API, but says they follow TC39 (Only for what they need), even tho Node has such module loader api
- They add `Deno.something` apis which are not TC39 compliant and never mentioned there but for this feature they do not do that
- In Deno is possible to load files which use `require` but you can't use `require.cache` since you are using mostly `import` which is not compatible with `require.cache` and doesn't show the imports in that array, If you try to use a module loader to customly load files and remove cache  (Wait there is not module loader :)
So this is a circle that just slows down the Deno adoption

Hope they accept this feature request and at least add some custom deno function for this https://github.com/denoland/deno/issues/25780

**Comment by petamoriken** [comment]
> * They make claims that everything node related is supported using some node layer, and you trust that and when you port your project you find this issue which has 4 years and are totally disappointed because you now are a victim of marketing, and you spent your time to port your project to deno even tho it doesn't fully work

I think that statement is incorrect. it wasn't neglected for 4 years, just that the experimental (unstable) API of Node.js is not implemented.
https://github.com/denoland/deno/tree/main/ext/node/polyfills#experimental

I understand the frustration of not getting the features you want, but if you want the features you want included, you can contribute in some way by discussing their usefulness and actually implementing them to Deno 🙂

**Comment by petamoriken** [comment]
For example, it's quite useful to report that the widely used npm package does not work properly due to the lack of implementation of the Node.js experimental Loader API.

**Comment by gherciu** [comment]
We know this is a experimental feature
But people come to this because `require.cache` doesn't work with `import` in Deno, and that would be an alternative to at least have a custom loader and somehow remove the `import` cache which is impossible at the moment in Deno.

If you check most of the issues above mentioned, are related to `hmr`, `require.cache`, `dynamic imports` and peoples get redirected to this issue and the ones with the actual bugs are closed by Deno team by mentioning this one for these 4 years, here is the actual issue, if we had the working `require.cache` with `import` then no one would even access this 4 years old issue till these days

**Comment by petamoriken** [comment]
Please calm down. If you want to have a problem with `require.cache`, you just provide a repository that shows that you can run it in Node.js but not in Deno.

> Sorry but without a full reproduction I can't really say what's going wrong. If you have a repo that can be tested then I can take a look.

https://github.com/denoland/deno/issues/25742#issuecomment-2364587853

**Comment by gherciu** [comment]
The issue which was closed has a clear description with reproduction steps, please re-read it

`How is that related to this issue`
In most of these issues you are redirected by Deno team to this issue, please re-read the issue and you'll see

**Comment by gherciu** [comment]
<img width="1053" alt="Screenshot 2024-11-03 at 17 59 58" src="https://github.com/user-attachments/assets/c527c1c3-5711-4cec-bfcc-4b16423a83a7">

**Comment by petamoriken** [comment]
Check the timeline. The issue was closed on September 20; a Deno team member requested a repository to reproduce your issue on September 21.

Again, could you please provide a repository to reproduce the issue?

**Comment by gherciu** [comment]
The reproduction was requested for another thing here https://github.com/denoland/deno/issues/25742#issuecomment-2364099717
Which is another bug in Deno, 
And yes `the issue was closed and then a reproduction was requested` which is not how things should work btw

**Comment by bradthomasbrown** [comment]
> you can contribute in some way by discussing their usefulness for compatibility or actually implementing them to Deno 🙂

The usefulness was discussed, in many issues. Those discussions were closed and redirected here. The only discussion here was minimized (read: censored). 
My impression is the exact opposite: no, nobody can contribute in some way by discussion and the Deno team has showed precisely why (no response + censoring).

I had plans to implement _something_ that could help but which would ultimately be a workaround, but seeing that discussion is censored by the Deno team if it is critical, but fair, to the Deno team drives me to not implement the feature in Deno.

The emoji usage comes off as passive aggressive, at least to me.

I also personally don't understand what Node.js or TC39 have to do with considering the functionality for Deno. It should not be required, for some feature, for it to be implemented somewhere else before it is considered for Deno. Otherwise, why use Deno?

**Comment by bartlomieju** [comment]
Could you folks list which packages you are trying to run that is blocked by not having access to loaders in Deno? Most recently we saw an issue with Playwright that uses TypeScript loader (that shouldn't really be necessary in Deno, but alas), but I'd like to gauge for a wider range of use cases before we undertake this.

Keep in mind that solving this issue will most likely result in worse performance for loading modules as Deno would have to check for registered loaders and potentially hopping multiple times between Rust and JavaScript code to actually load a file before handing it off to V8 for actual execution.

**Comment by gherciu** [comment]
In my case, we build a new package which will need to load user code, and this code may change at the runtime, and since `require.cache` doesn't work correctly in Deno with `import` (you can't use even `require` inside a file that uses `import`) we can't refresh the cache at runtime (Which is a requirement, and we can't stop the deno process, because it will close server connections)
In deno you can load a different version using `?query=param` but it just loads the first level but the children import again it takes from cache

This package will be used by a big tech company (I can't say the name due to the contract)

For example in `Bun` the `require.cache` also shows the modules loaded with `import` (and you can use these both in same file and no Module related error is thrown) and that is a option we think now, maybe to move to Bun, if in Deno there is no way to achieve it, but Bun has at the moment a small issue to fix, it doesn't return the children imports of a module and is hard to figure out which children's also needs to be deleted from cache and we wait for that

**Comment by petamoriken** [comment]
Since [caching is defined in the language specification](https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#sec-cyclic-module-records:~:text=Evaluation%20must%20be%20only%20performed%20once%2C%20as%20it%20can%20cause%20side%20effects) when using ES Modules (Cyclic Module Records), this seems to be the only way to accomplish this in CommonJS (Abstract Module Records).

> For example in `Bun` the `require.cache` also shows the modules loaded with `import` (and you can use these both in same file and no Module related error is thrown) and that is a option we think now

My understanding is that Bun's implementation deviates from JavaScript...

---

By the way, this issue could be solved by implementing `require(esm)` feature in Node.js v23 to Deno. That feature seems to work [without a flag](https://github.com/nodejs/node/pull/55085), but it is [still experimental](https://nodejs.org/docs/v23.1.0/api/modules.html#loading-ecmascript-modules-using-require).

I confirmed that the `require.cache` contains the value of the module:

<img width="931" alt="use require(esm) in Node.js v23" src="https://github.com/user-attachments/assets/8243bd94-e739-49f4-98fd-7b6392344020">

**Comment by gherciu** [comment]
At the moment in Deno, `import` and `require` just live separate lives.
You can't interact between these 2.

In node you at least can compile by yourself `.ts` to `.js` and transform all imports to `require` and you have at the runtime only one wy of importing things and you can remove the cache

But in deno there is no compile step and everything stays as `import` or `require` and you just can't uncache something that used `import`, like literally there is no way.

**Comment by bartlomieju** [comment]
> At the moment in Deno, import and require just live separate lives.
You can't interact between these 2.

You can interact between them just like in Node.js.

> In node you at least can compile by yourself .ts to .js and transform all imports to require and you have at the runtime only one way of importing things and you can remove the cache. In Bun this is partially implemented even tho they have a bug at the moment. But they have an alternative to that which is Bun.plugin you may use it as a loader which is an alternative to Module.loader



You can compile these files yourself in Deno as well and you can `require` them as well. You can only remove from cache if you call `require()`, you can't remove from cache if you `import()` something, neither in Node, nor Deno. I can't speak about Bun, because I haven't used that API.

Let's see a short example:
```js
// main.cjs
(async () => {
    const fs = require("fs");
    fs.writeFileSync("./import.mjs", "export default { foo: 'bar' };");
    const foo = await import("./import.mjs");
    console.log(foo);

    console.log(require.cache);
    Object.keys(require.cache).forEach(key => delete require.cache[key]);
    console.log(require.cache);
    
    fs.writeFileSync("./import.mjs", "export default { foo: 'fizz' };");
    const foo1 = await import("./import.mjs");
    console.log(foo1);
})();
```

```
$ node main.cjs
[Module: null prototype] { default: { foo: 'bar' } }
[Object: null prototype] {
  '/Users/ib/dev/scratch_node_import/main.cjs': {
    id: '.',
    path: '/Users/ib/dev/scratch_node_import',
    exports: {},
    filename: '/Users/ib/dev/scratch_node_import/main.cjs',
    loaded: true,
    children: [],
    paths: [
      '/Users/ib/dev/scratch_node_import/node_modules',
      '/Users/ib/dev/node_modules',
      '/Users/ib/node_modules',
      '/Users/node_modules',
      '/node_modules'
    ],
    [Symbol(kIsMainSymbol)]: true,
    [Symbol(kIsCachedByESMLoader)]: false,
    [Symbol(kIsExecuting)]: false
  }
}
[Object: null prototype] {}
[Module: null prototype] { default: { foo: 'bar' } }
```

```
$ deno -RW main.cjs
[Module: null prototype] { default: { foo: "bar" } }
[Object: null prototype] {
  "/Users/ib/dev/scratch_node_import/main.cjs": Module {
    id: ".",
    path: "/Users/ib/dev/scratch_node_import",
    exports: {},
    filename: "/Users/ib/dev/scratch_node_import/main.cjs",
    loaded: true,
    parent: null,
    children: [],
    paths: [
      "/Users/ib/dev/scratch_node_import/node_modules",
      "/Users/ib/dev/node_modules",
      "/Users/ib/node_modules",
      "/Users/node_modules",
      "/node_modules"
    ]
  }
}
[Object: null prototype] {}
[Module: null prototype] { default: { foo: "bar" } }
```

> But in Deno there is no compile step and everything stays as import or require and you just can't uncache something that used import, like literally there is no way. You can't use require.cache with import and the Loader alternative proposed is not even implemented.

You can have a dedicated compilation/transpilation step like a lot of frameworks do (Vite, Next.js, etc). It's the transpilation output that determines if file uses `require()` (ie. CommonJS format) or `import()` (ie. ES modules format). 

Altering `require.cache` doesn't have impact on `import` in Node either - as per the ES spec, once an ES module is loaded it can't be unloaded. 

While I can see why loaders might be a useful feature, there's a whole ecosystem that worked without loaders by handling transpilation it self. It is expected that it will work exactly the same in Deno as it does in Node.js.

If you have some example code that works in Node, but has wrong behavior in Deno I'll be glad to take a look and fix the bug.

> By the way, this issue could be solved by implementing require(esm) feature in Node.js v23 to Deno. That feature seems to work https://github.com/nodejs/node/pull/55085, but it is [still experimental](https://nodejs.org/docs/v23.1.0/api/modules.html#loading-ecmascript-modules-using-require).

Deno supports `require(ESM)` as of Deno v2.0.

**Comment by petamoriken** [comment]
> Deno supports require(ESM) as of Deno v2.0.

I didn't know that. Thanks!

**Comment by gherciu** [comment]
`You can compile these files yourself in Deno as well and you can require them as well. `

I mean, why then use Deno? If in the end I can run compiled files with Node, that doesn't make sense
The whole point of not transpiling files and running TS as is,  just gone then

So the idea is just to move back to node and do the compilation, that's what I understand

And that's the problem, we apply a Node mindset to Deno mindset

**Comment by CyanChanges** [comment]
> Could you folks list which packages you are trying to run that is blocked by not having access to loaders in Deno? Most recently we saw an issue with Playwright that uses TypeScript loader (that shouldn't really be necessary in Deno, but alas), but I'd like to gauge for a wider range of use cases before we undertake this.
> 
> Keep in mind that solving this issue will most likely result in worse performance for loading modules as Deno would have to check for registered loaders and potentially hopping multiple times between Rust and JavaScript code to actually load a file before handing it off to V8 for actual execution.

In @cordisjs/loader (https://github.com/cordiverse/cordis),
when a hot module reload triggered
https://github.com/cordiverse/cordis/blob/master/packages/hmr/src/index.ts#L252-L254 
(`this.internal` is `esmLoader` from `node:internal/modules/esm/loader`)
the cache of the imported plugin and all related services is deleted from loadCache,
after that, we imported the plugin again and plug the plugin back.
Without accessing the loader, we can't invalidate the loadCache,
even we can reload the entrypoint, the children imports of the plugin can't be reloaded.

**Comment by dc0d** [comment]
This idea (thought experiment) may seem unusual, but it could open the door to some pragmatic solutions.

What if Deno offered a mechanism to “inline” another module directly into the current one? Instead of a standard import, a special directive could copy the referenced module’s code right into the importing file. 

```typescript
// file1.ts

export const moduleLevelState = { state: 'moduleLevelState' };
export const act1 = () => {};
export const act2 = () => {};
``` 

```typescript
// file2.ts

import inline('./file1.ts');

// overwrites the definition for act2
export const act2 = () => {};

// and duplicates/copies the definition for moduleLevelState & act1
// so moduleLevelState in file1.ts is not related to moduleLevelState in file2.ts
```

It seems in most cases people need to reload a module a specific number of times. If this assumption is backed by reality, inlining modules could be a valid option - that breaks the least number of things (for example deno's security system should not face many problems while handling this).

**Comment by CyanChanges** [comment]
We should have something like this
```typescript
// mod1.ts
console.log("Hello World")
```
```typescript
// user.ts
await import("./mod1.ts")
await import("./mod1.ts")

console.log("delete the cache")
delete Module.Loader.cache["./mod1.ts"]

await import("./mod1.ts")
```
```shell
$ deno run user.ts
Hello World
delete the cache
Hello World
```

Thinking about that `user.ts` is created by the user, and you can't change it (like add `#<random-number>` to do a uncached import).

**Comment by CyanChanges** [comment]
I have created a fork of Deno (which I just decided to called it Done), implements a very simple loader api

here is the demonstrate of invalidating import cache

![Image](https://github.com/user-attachments/assets/a3620454-7349-4fd3-8652-c44f3ab514d4)

**Comment by cyans-nya** [comment]
I have making progress by try provide the ability to mutate the `ModuleMap` (a struct in deno_core that handles all imports including reuse evaluated module (module cache) and request actual `dyn ModuleLoader` for loading)
If you are willing to push this forward, please checkout the issue and the pr, and show your opinion about this 

I will update the progress of this here.

**Progress**
- [ ] deno_core: feat: enhance ModuleMap https://github.com/denoland/deno_core/issues/1143
       https://github.com/denoland/deno_core/issues/1146
- [ ] deno: add ext/loader to provide `Deno.loader` to users
- [ ] deno: maybe also import hooks support?

We could put this in Deno under a permission flag or unstable flag, require it to be explicit enabled by the user. 
This enable framework developer to have HMR implemented without pre-processing user code.

### This could a possible spec violation? 

> Since caching is defined in the language specification when using ES Modules (Cyclic Module Records), this seems to be the only way to accomplish this in CommonJS (Abstract Module Records).
> > [Evaluation](https://tc39.es/ecma262/multipage/syntax-directed-operations.html#sec-evaluation) must be only performed once, as it can cause side effects; it is thus important to remember whether evaluation has already been performed, even if unsuccessfully. (In the error case, it makes sense to also remember the exception because otherwise subsequent Evaluate() calls would have to synthesize a new one.)

Firstly, the module would not be re-evaluated (i guess it's not possible to re-evaluate a v8::Module ?)

So, what we are doing here is not re-evaluating a existing module, 
but we are creating a new module, and reassign a existing specifier to a **new module**'s id, 

So the old module is still there, would not be removed, or evaluated again.
we are only making the specifier resolves to the new module.

So, I don't think this is a spec violation. As we are not touching the original Module by anyway.

**Comment by Hajime-san** [comment]
My understanding is that there is no API to clear a module in V8, so this will be a non-ESM, CJS-like immediately evaluated object.
I think it would be more constructive to create parallel PRs to deno(this repository) and discuss them, rather than[ discussing the API design in an issue](https://github.com/denoland/deno_core/issues/1143) due to `deno_core` is not a public interface.

**Comment by cyans-nya** [comment]
> My understanding is that there is no API to clear a module in V8, so this will be a non-ESM, CJS-like immediately evaluated object. I think it would be more constructive to create parallel PRs to deno(this repository) and discuss them, rather than[ discussing the API design in an issue](https://github.com/denoland/deno_core/issues/1143) due to `deno_core` is not a public interface.

I would consider create a PR to deno too.
If you're willing to see, I have a prototype that demonstrates the Loader API in https://github.com/CyanProjects/done/tree/main/ext/loader

**Comment by iacore** [comment]
API Proposal from Lume: https://github.com/lumeland/lume/issues/590#issuecomment-3008927969

```
addEventListener("filechanged", (ev) => {
  const { filename, canHotReload } = ev.details;
  if (canHotReload) {
    console.log(`The file ${filename} was changed and can be hot reloaded`);
    Deno.hotReload(filename); // Hot reload the file
  } else {
    console.log(`The file ${filename} was changed and cannot be hot reloaded. Restarting process...`);
    Deno.restart();
  }
});
```

## Our Use Case

Lume is a static site generator. A page may depend on Deno to execute a Typescript file. When that file changes, we hope to 

1. reload that file
2. render the page depending on that file

**Comment by abflow** [comment]
I would personally enjoy creating ES modules on the fly via a Module Loader API conforming to the module graph, with full recursivity.

Example: The module loader API could intercept predefined path patterns, retrieve data from a Deno.Kv store, transform it as an ES Module and then incorporate it in the Deno module graph the same way as file-based ES Modules.

Retrieval and transformation steps would be user-defined.

**Comment by Hajime-san** [comment]
- https://github.com/nodejs/node/pull/61767

There is a PR for Node.js to clear the ESM cache. It seems like this is going in the direction of allowing memory leaks since it is not possible to clear modules loaded in Realm.

**Comment by bartlomieju** [comment]
Done in https://github.com/denoland/deno/pull/33763 and https://github.com/denoland/deno/pull/33849. We're not gonna do a bespoke API, the one from Node should work just fine.
